File: minimumSAROAT.ts
function minimumScore(nums: number[], edges: number[][]): number {
    const n = nums.length;
    const graph: number[][] = Array.from({ length: n }, () => []);
    for (const [u, v] of edges) {
        graph[u].push(v);
        graph[v].push(u);
    }

    const xor: number[] = Array(n).fill(0);
    const tin: number[] = Array(n).fill(0);
    const tout: number[] = Array(n).fill(0);
    let time = 0;

    function dfs(u: number, parent: number): number {
        tin[u] = time++;
        xor[u] = nums[u];
        for (const v of graph[u]) {
            if (v !== parent) {
                xor[u] ^= dfs(v, u);
            }
        }
        tout[u] = time++;
        return xor[u];
    }

    dfs(0, -1);
    const total = xor[0];

    function isAncestor(u: number, v: number): boolean {
        return tin[u] <= tin[v] && tout[v] <= tout[u];
    }

    const edgeNodes: number[] = [];
    for (const [u, v] of edges) {
        edgeNodes.push(isAncestor(u, v) ? v : u);
    }

    let minScore = Infinity;

    for (let i = 0; i < edgeNodes.length; i++) {
        const a = edgeNodes[i];
        for (let j = i + 1; j < edgeNodes.length; j++) {
            const b = edgeNodes[j];
            let x1: number, x2: number, x3: number;

            if (isAncestor(a, b)) {
                x1 = xor[b];
                x2 = xor[a] ^ xor[b];
                x3 = total ^ xor[a];
            } else if (isAncestor(b, a)) {
                x1 = xor[a];
                x2 = xor[b] ^ xor[a];
                x3 = total ^ xor[b];
            } else {
                x1 = xor[a];
                x2 = xor[b];
                x3 = total ^ xor[a] ^ xor[b];
            }

            const vals = [x1, x2, x3];
            minScore = Math.min(minScore, Math.max(...vals) - Math.min(...vals));
        }
    }

    return minScore;
}

File: minimumSAROAT.rs
use std::collections::HashSet;

impl Solution {
    pub fn minimum_score(nums: Vec<i32>, edges: Vec<Vec<i32>>) -> i32 {
        let n = nums.len();
        let mut graph = vec![Vec::new(); n];
        for e in edges.iter() {
            let (u, v) = (e[0] as usize, e[1] as usize);
            graph[u].push(v);
            graph[v].push(u);
        }

        let mut xor = vec![0; n];
        let mut in_time = vec![0; n];
        let mut out_time = vec![0; n];
        let mut time = 0;

        fn dfs(u: usize, parent: usize, graph: &Vec<Vec<usize>>, nums: &Vec<i32>,
               xor: &mut Vec<i32>, in_time: &mut Vec<usize>, out_time: &mut Vec<usize>, time: &mut usize) {
            in_time[u] = *time;
            *time += 1;
            xor[u] = nums[u];
            for &v in &graph[u] {
                if v != parent {
                    dfs(v, u, graph, nums, xor, in_time, out_time, time);
                    xor[u] ^= xor[v];
                }
            }
            out_time[u] = *time;
        }

        dfs(0, n, &graph, &nums, &mut xor, &mut in_time, &mut out_time, &mut time);
        let total = xor[0];
        let mut answer = i32::MAX;

        for i in 1..n {
            for j in 1..n {
                if i == j { continue; }

                let (a, b, c) = if in_time[i] < in_time[j] && out_time[j] <= out_time[i] {
                    let x = xor[j];
                    let y = xor[i] ^ xor[j];
                    let z = total ^ xor[i];
                    (x, y, z)
                } else if in_time[j] < in_time[i] && out_time[i] <= out_time[j] {
                    let x = xor[i];
                    let y = xor[j] ^ xor[i];
                    let z = total ^ xor[j];
                    (x, y, z)
                } else {
                    let x = xor[i];
                    let y = xor[j];
                    let z = total ^ x ^ y;
                    (x, y, z)
                };

                let min_val = *[a, b, c].iter().min().unwrap();
                let max_val = *[a, b, c].iter().max().unwrap();
                answer = answer.min(max_val - min_val);
            }
        }

        answer
    }
}
